### DISCUSSION TITLE
[BUG] The 'Connect-PnPOnline' command was found in the module 'PnP.PowerShell', but the module could not be loaded.

### DISCUSSION THREAD

**Discussion by myatix** [pnp/powershell#3424] [Q&A]
Hi All,

I am experiencing the below error in Azure Function Apps, can someone else tell me if they have the same issue or perhaps point me to a solution.

### Reporting an Issue or Missing Feature

**ERROR:** The 'Connect-PnPOnline' command was found in the module 'PnP.PowerShell', but the module could not be loaded. For more information, run 'Import-Module PnP.PowerShell'.

**Requirements.psd1**
```
# This file enables modules to be automatically managed by the Functions service.
# See https://aka.ms/functionsmanageddependency for additional information.
#
@{
    # For latest supported version, go to 'https://www.powershellgallery.com/packages/Az'.
    # To use the Az module in your function app, please uncomment the line below.
    #'PnP.PowerShell' = '1.12.0'
    'Az' = '10.*'
    'PnP.PowerShell' = '2.1.1'
    
}
```

**Profile.ps1**
```
# Azure Functions profile.ps1
#
# This profile.ps1 will get executed every "cold start" of your Function App.
# "cold start" occurs when:
#
# * A Function App starts up for the very first time
# * A Function App starts up after being de-allocated due to inactivity
# Authenticate with Azure PowerShell using MSI.
# Remove this if you are not planning on using MSI or Azure PowerShell.
#if ($env:MSI_SECRET) {
#    Disable-AzContextAutosave -Scope Process | Out-Null
#    Connect-AzAccount -Identity
#}

#Import Modules into Function App
Import-Module Az
Import-Module PnP.PowerShell

```
**Azure Function Config**
PowerShell Version: 7.0
Runtime version: 4.25.2.21138

### Expected behavior 
Please describe what output you expect to see from the PnP PowerShell Cmdlets

### Actual behavior 
Please describe what you see instead. Please provide samples of output or screenshots.

### Steps to reproduce behavior

```
using namespace System.Net
param($Request)

$env:PNPPOWERSHELL_UPDATECHECK="false"
$tenant = $env:Tenant
$clientId = $env:ClientId
$certificateBase64Encoded = $env:Certificate
$templateLibUrl = $env:TemplateLibUrl
$templateFileName = $env:TemplateFileName
$teamSiteLogo = "logo.jpg"
$siteDirListTitle = "Site Directory"
$downloadFilePath = $PSScriptRoot
$webUrl = $Request.Body.webUrl

write-output "Starting configuration of Modern Teams Site:"
write-output $webUrl

Import-Module PnP.PowerShell

$webConnect = Connect-PnPOnline -Url $webUrl -Tenant $tenant -ClientId $clientId -CertificateBase64Encoded $certificateBase64Encoded -ReturnConnection

```

### What is the version of the Cmdlet module you are running?
See above...

### Which operating system/environment are you running PnP PowerShell on?
- [x ] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by himani-chi** [comment]
Why has this issue been converted to discussion? It's happening for me as well.

**Comment by myatix** [comment]
After hours of testing and no help from the PnP.Powershell community I discovered that if you are using Managed Identity for authentification you need to have the following uncommented.

It would appear that the PnP article about setting up an Azure function is incorrect when using Managed Identities and using a certificate currently doesn't appear to work with PnP.Powershell 2.2.x and PowerShell 7.2.

https://pnp.github.io/powershell/articles/azurefunctions.html

![image](https://github.com/pnp/powershell/assets/48433526/cfd5b4f6-699b-44b1-8804-91aa9fe98991)

```
# Azure Functions profile.ps1
#
# This profile.ps1 will get executed every "cold start" of your Function App.
# "cold start" occurs when:
#
# * A Function App starts up for the very first time
# * A Function App starts up after being de-allocated due to inactivity
# Authenticate with Azure PowerShell using MSI.
# Remove this if you are not planning on using MSI or Azure PowerShell.
 Authenticate with Azure PowerShell using MSI.
# Remove this if you are not planning on using MSI or Azure PowerShell.
if ($env:MSI_SECRET) {
    Disable-AzContextAutosave -Scope Process | Out-Null
    Connect-AzAccount -Identity
}

#Import Modules into Function App
Import-Module -Name Az.Accounts
Import-Module -Name PnP.PowerShell
```

This got rid of the issue for me...

requirements.psd1
```
# This file enables modules to be automatically managed by the Functions service.
# See https://aka.ms/functionsmanageddependency for additional information.
#
@{
    # For latest supported version, go to 'https://www.powershellgallery.com/packages/Az'.
    # To use the Az module in your function app, please uncomment the line below.
    #'PnP.PowerShell' = '1.12.0'
    'Az.Accounts' = '2.2.3'
    'PnP.PowerShell' = '2.2.74-nightly'
    
}
```

**Comment by AndersRask** [comment]
> After hours of testing and _no help from the PnP.Powershell community_...
> 

PnP.PowerShell is a community driven module that people are contributing to or helping with in their spare time. You cannot expect any of us to drop everything we are doing just because you have an issue.
If you think you've found a solution, feel free to contribute by suggesting an edit to the documentation :-)

Searching the comments here you will find lots of issues with assembly collisions and module load issues. Most of them have been bypassed in the more recent builds, so it is often a good idea to try the latest builds if you have issues of that nature.

**Comment by myatix** [comment]
@AndersRask  I am sorry if I offended you... I am just use to chatting in more active communities I guess.

It's really difficult to learn how to use this stuff when the documentation appears to be incorrect and doing basic things like just connecting causes so many issues. I totally get the value running pnp.powershell on a client but it is incredibly temperamental in Azure Functions.

But thanks again...

**Reply by AndersRask** [reply]
I am not offended. Just making clear that this is very busy people doing this in their spare time, so set your expectations accordingly :)

The modules are evolving slowly but surely, the assembly issue was solved recently, so if this is related to that, this could be the cause. Again, if you have improvements to the documentation or the code, it is all a part of this repository, so feel free to contribute :)
